google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher2.java

```diff
@@ -0,0 +1,284 @@
+    this.systemExecutor = systemExecutor;
+    this.receiver = receiver;
+    this.flowController = flowController;
+  }
+
+  synchronized void start(Collection<? extends Connection> connections) {
+    this.connections.addAll(connections);
+    this.ackNackJob =
+        systemExecutor.scheduleAtFixedRate(
+            new Runnable() {
+              @Override
+              public void run() {
+                StreamingPullRequest.Builder builder = StreamingPullRequest.newBuilder();
+                boolean more = true;
+                while (more) {
+                  builder.clear();
+                  more = drainAckNacks(builder);
+                  // If both are 0, there's nothing to send.
+                  if (builder.getAckIdsCount() != 0
+                      || builder.getModifyDeadlineAckIdsCount() != 0) {
+                    sendRequest(builder.build());
+                  }
+                }
+              }
+            },
+            100,
+            100,
+            TimeUnit.MILLISECONDS);
+
+    this.extensionJob =
+        systemExecutor.scheduleAtFixedRate(
+            new Runnable() {
+              @Override
+              public void run() {
+                List<String> extendIds = new ArrayList<>(extensionSet.keySet());
+                for (List<String> batch : Lists.partition(extendIds, MAX_CHANGE_PER_REQUEST)) {
+                  sendRequest(
+                      StreamingPullRequest.newBuilder()
+                          .addAllModifyDeadlineAckIds(batch)
+                          .addAllModifyDeadlineSeconds(
+                              Collections.nCopies(batch.size(), DEADLINE_EXTENSION_SEC))
+                          .build());
+                }
+              }
+            },
+            KEEPALIVE_SEC,
+            KEEPALIVE_SEC,
+            TimeUnit.SECONDS);
+  }
+
+  synchronized void shutdown() {
+    ackNackJob.cancel(false);
+    extensionJob.cancel(false);
+  }
+
+  private boolean drainAckNacks(StreamingPullRequest.Builder builder) {
+    while (builder.getAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      String id = idsToAck.poll();
+      if (id == null) {
+        break;
+      }
+      builder.addAckIds(id);
+    }
+
+    while (builder.getModifyDeadlineAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      ModAckItem modAck = modAcks.poll();
+      if (modAck == null) {
+        break;
+      }
+      builder.addModifyDeadlineAckIds(modAck.ackId());
+      builder.addModifyDeadlineSeconds(modAck.seconds());
+    }
+
+    return builder.getAckIdsCount() == MAX_CHANGE_PER_REQUEST
+        || builder.getModifyDeadlineAckIdsCount() == MAX_CHANGE_PER_REQUEST;
+  }
+
+  void processReceivedMessages(List<ReceivedMessage> messages, Runnable doneCallback) {
+    if (messages.isEmpty()) {
+      doneCallback.run();
+      return;
+    }
+
+    for (ReceivedMessage message : messages) {
+      modAcks.add(ModAckItem.create(message.getAckId(), DEADLINE_EXTENSION_SEC));
```

Comment:
if we process messages fast, we could end up both modacking and acking the same message in one request. I don't think that's a problem right?

Reply:
That's fine